### PR TITLE
feat: add payment capture form for orders

### LINKED ISSUE #5064: feat: add payment capture form for orders
Resolving : https://github.com/juspay/hyperswitch-control-center-cloud/issues/172

### PR DESCRIPTION
**Pull request by Riddhiagrawal001**
## Type of Change

- [x] New feature

## Description

Adds an `OrderCaptureForm` that lets users capture (full or partial) payments directly from the order details screen.

- **APIUtils**: Add `PAYMENT_CAPTURE` entity mapping to `payments/{payment_id}/capture` (POST).
- **HSwitchOrderUtils**: Parameterize `amountFieldWithPrecision` with `~name`, `~label`, and `~placeholder` (defaulting to the existing refund values) so the field can be reused for the capture amount input.
- **OrderCaptureForm.res** (new): Capture form with amount pre-filled to `amount_capturable`, client-side validation (must be > 0 and not exceed the capturable amount), and submit wiring to the capture endpoint with success/error toasts.
- **Status handling**: Add the `RequiresCapture` status across `OrderEntity`, `HSwitchOrderUtils`, and `PaymentIntentEntity` (status variant mapping + label styling).

## Motivation and Context

Payments in a `requires_capture` state previously had no way to be captured from the dashboard. This PR adds a dedicated capture form so users can capture the full or a partial amount directly from the order details view, and ensures the `RequiresCapture` status renders correctly wherever payment status is displayed.

## How did you test it?

Tested manually:
- Opened an order in `RequiresCapture` state and captured the full amount.
- Captured a partial amount and confirmed validation rejects amounts above the capturable value and amounts ≤ 0.
- Verified success and error toasts, and that the list refetches after a successful capture.

## Where to test it?

- [ ] INTEG
- [x] SANDBOX
- [ ] PROD

## Backend Dependency

- [x] No

Backend PR URL:

## Feature Flag

- [x] No feature flag changes

Feature flag name(s):

## Checklist

- [x] I ran `npm run re:build`
- [x] I reviewed submitted code
- [ ] I added unit tests for my changes where possible

### CODE REVIEW

File: src/screens/Transaction/Order/OrderCaptureForm.res [hunk 1]
```diff
@@ -0,0 +1,164 @@
...
+      setIsLoading(_ => true)
+      let captureUrl = getURL(
+        ~entityName=V1(PAYMENT_CAPTURE),
+        ~methodType=Post,
+        ~id=Some(order.payment_id),
+      )
+
+      let dict = values->getDictFromJsonObject
+      let amount = dict->getFloat("amount", 0.0)
+      let body =
```
Comment (XyneSpaces) [inline src/screens/Transaction/Order/OrderCaptureForm.res:42]: **[S2] Critical Issue**: The catch block silently discards all errors with `_`. This makes debugging production issues nearly impossible. Please log the error before showing the toast:

```rescript
| exn => {
    Console.error2("Payment capture failed:", exn)
    showToast(~message="Failed to capture payment", ~toastType=ToastError)
    setIsLoading(_ => false)
    setShowModal(_ => false)
  }
```

Additionally, `setIsLoading(_ => false)` is missing in the error case, leaving the UI stuck in a loading state.

File: src/screens/Transaction/Order/OrderCaptureForm.res [hunk 2]
```diff
@@ -0,0 +1,164 @@
...
+  // Full amount available to capture, in major units, used as the pre-filled value
+  let amountCapturableInMajorUnits = order.amount_capturable /. conversionFactor
+  let amountToCaptureField = amountFieldWithPrecision(
+    ~precisionDigits,
+    ~label="Amount to Capture",
+    ~placeholder="Enter Amount to Capture",
+  )
+
+  let initiateValueJson =
+    [("amount", amountCapturableInMajorUnits->JSON.Encode.float)]
```
Comment (XyneSpaces) [inline src/screens/Transaction/Order/OrderCaptureForm.res]: **[S1] High Priority**: Missing `~name` parameter in `amountFieldWithPrecision` call. While the function has default values, explicitly passing `~name="amount"` would make the code more readable and ensure the form field name matches the validation logic.

**Review by XyneSpaces — COMMENTED** [review]
## Automated Review Summary

This PR adds payment capture functionality to the Control Center. Overall, the implementation looks clean and follows established patterns. However, I've identified a few areas for improvement:

### Key Findings

1. **Critical**: Error handling in `capturePayment` catches all exceptions silently - the `catch` block discards the error without logging it, making debugging difficult.

2. **High**: Loading state is not reset when capture fails - `setIsLoading(_ => false)` is missing in the error handler.

3. **Medium**: The `amountFieldWithPrecision` function parameterization is good but could benefit from a `~name` parameter documentation.

Please address the critical issues before merging.

File: src/screens/Transaction/Order/OrderCaptureForm.res [hunk 3]
```diff
@@ -0,0 +1,164 @@
...
+        ~methodType=Post,
+        ~id=Some(order.payment_id),
+      )
+
+      let dict = values->getDictFromJsonObject
+      let amount = dict->getFloat("amount", 0.0)
+      let body =
+        [("amount_to_capture", Math.round(amount *. conversionFactor)->JSON.Encode.float)]
+        ->getJsonFromArrayOfJson
+
```
Comment (XyneSpaces) [inline src/screens/Transaction/Order/OrderCaptureForm.res:46]: **🟡 Minor Issue**: Error handling silently swallows exceptions. Consider logging the error details for debugging:
```rescript
| exn => {
    Console.error2("Payment capture failed", exn)
    showToast(~message="Failed to capture payment", ~toastType=ToastError)
    setShowModal(_ => false)
  }
```

File: src/screens/Transaction/Order/OrderCaptureForm.res [hunk 4]
```diff
@@ -0,0 +1,164 @@
...
+    }
+  }
+
+  let onSubmit = (values, _) => {
+    open Promise
+    capturePayment(values)->ignore
+    Nullable.null->resolve
+  }
+
+  let validate = values => {
```
Comment (XyneSpaces) [inline src/screens/Transaction/Order/OrderCaptureForm.res]: **🟡 Suggestion**: The floating-point comparison here could have precision issues. Consider rounding both values to integers before comparison for more reliable validation.

File: src/screens/Transaction/Order/OrderCaptureForm.res [hunk 5]
```diff
@@ -0,0 +1,164 @@
+open OrderEntity
```
Comment (XyneSpaces) [inline src/screens/Transaction/Order/OrderCaptureForm.res:1]: **✅ Good**: The new component follows the existing patterns for forms and API integration. The validation logic is well-structured with clear error messages.

**Review by XyneSpaces — COMMENTED** [review]
## Code Review: Payment Capture Form

This PR adds an `OrderCaptureForm` for capturing payments in `requires_capture` state. Here are my findings:

### Issues Found:

1. **Error handling in capturePayment (Line 47-50)**: The catch block silently swallows all errors. Consider logging the error details for debugging purposes.

2. **Amount precision handling**: The validation logic compares amounts after rounding, which could lead to edge cases with floating-point precision. Consider using integer arithmetic for comparisons.

3. **Missing tests**: PR checklist indicates unit tests were not added. Consider adding tests for the validation logic and the capture flow.

### Suggestions:

1. The `amountFieldWithPrecision` function parameterization is a good refactoring that enables reuse.
2. Status handling for `RequiresCapture` is consistently applied across entities.

Overall, the implementation looks solid but would benefit from error logging and test coverage.

**Comment by XyneSpaces** [comment]
## 📋 PR Review Summary

### Review Status: ✅ Approved with Suggestions

**Key Changes:**
- Added `OrderCaptureForm` for capturing payments in `requires_capture` state
- Parameterized `amountFieldWithPrecision` for reusability
- Added `RequiresCapture` status across Order entities

**Findings:**
| Severity | Count | Description |
|----------|-------|-------------|
| 🟡 Minor | 2 | Error logging, floating-point comparison |
| ✅ Positive | 3 | Good patterns, consistent status handling |

**Action Items:**
1. Add error logging in catch block for debugging
2. Consider integer arithmetic for amount comparisons
3. Add unit tests for validation logic

Overall solid implementation following existing patterns.

**Comment by XyneSpaces** [comment]
Fetching comments for testing

**Comment by XyneSpaces** [comment]
temp

File: src/screens/Transaction/Order/OrderCaptureForm.res [hunk 6]
```diff
@@ -0,0 +1,163 @@
...
+
+  let validate = values => {
+    let errors = Dict.make()
+    let valuesDict = values->getDictFromJsonObject
+    switch valuesDict->getOptionFloat("amount") {
+    | Some(floatVal) =>
+      let enteredAmountInMinorUnits = Math.round(floatVal *. conversionFactor)
+      let capturableInMinorUnits = Math.round(amountCapturableInMajorUnits *. conversionFactor)
+      if enteredAmountInMinorUnits > capturableInMinorUnits {
+        let formattedAmount = Float.toFixedWithPrecision(
```
Comment (susrithasabbini) [inline src/screens/Transaction/Order/OrderCaptureForm.res]: Please move to utils file

File: src/screens/Transaction/Order/OrderCaptureForm.res [hunk 7]
```diff
@@ -0,0 +1,163 @@
...
+  // Full amount available to capture, in major units, used as the pre-filled value
+  let amountCapturableInMajorUnits = order.amount_capturable /. conversionFactor
+  let amountToCaptureField = amountFieldWithPrecision(
+    ~precisionDigits,
+    ~label="Amount to Capture",
+    ~placeholder="Enter Amount to Capture",
+  )
+
+  let initiateValueJson =
+    [("amount", amountCapturableInMajorUnits->JSON.Encode.float)]->getJsonFromArrayOfJson
```
Comment (susrithasabbini) [inline src/screens/Transaction/Order/OrderCaptureForm.res]: Please move to utils

File: src/screens/Transaction/Order/OrderCaptureForm.res [hunk 8]
```diff
@@ -0,0 +1,163 @@
...
+    | None => Dict.set(errors, "amount", "Required"->JSON.Encode.string)
+    }
+    errors->JSON.Encode.object
+  }
+
+  <div>
+    <Form onSubmit validate initialValues=initiateValueJson>
+      <div className="flex flex-col w-full max-w-4xl mx-auto p-6">
+        <div
+          className="border-b border-jp-gray-940 border-opacity-75 dark:border-jp-gray-960 dark:border-opacity-75 pb-4 mb-6">
```
Comment (susrithasabbini) [inline src/screens/Transaction/Order/OrderCaptureForm.res]: Please use new nd gray colors

File: src/screens/Transaction/Order/ShowOrder.res [hunk 9]
```diff
@@ -469,6 +487,15 @@ module OrderActions = {
           refetch
         />
       </Modal>
+      <Modal
+        showModal=showCaptureModal
+        setShowModal=setShowCaptureModal
+        borderBottom=true
+        childClass=""
+        modalClass="w-full md:w-4/12 mx-auto mt-20"
+        bgClass="bg-white dark:bg-jp-gray-darkgray_background">
```
Comment (susrithasabbini) [inline src/screens/Transaction/Order/ShowOrder.res]: New nd gray colors

**Comment by muditbhutani** [comment]
What happens on failure(connector not supporting this flow)?

**Review by susrithasabbini — DISMISSED** [review]

File: src/screens/Transaction/Order/OrderCaptureForm.res [hunk 10]
```diff
@@ -0,0 +1,141 @@
...
+      let dict = values->getDictFromJsonObject
+      let amount = dict->getFloat("amount", 0.0)
+      let body =
+        [
+          ("amount_to_capture", Math.round(amount *. conversionFactor)->JSON.Encode.float),
+        ]->getJsonFromArrayOfJson
+
+      let _ = await updateDetails(captureUrl, body, Post)
+      refetch()->ignore
+      showToast(~message="Payment captured successfully", ~toastType=ToastSuccess)
```
Comment (kanikabansal08) [inline src/screens/Transaction/Order/OrderCaptureForm.res:49]: When a capture is declined does HS return a non 2xx error or does it return 200 with status "failed" in the body?

File: src/screens/Transaction/Order/OrderCaptureForm.res [hunk 11]
```diff
@@ -0,0 +1,141 @@
...
+  let precisionDigits = getAmountPrecisionDigits(order.currency)
+  // Full amount available to capture, in major units, used as the pre-filled value
+  let amountCapturableInMajorUnits = order.amount_capturable /. conversionFactor
+  let amountToCaptureField = amountFieldWithPrecision(
+    ~precisionDigits,
+    ~label="Amount to Capture",
+    ~placeholder="Enter Amount to Capture",
+  )
+
+  let initiateValueJson = getCaptureInitialValues(~amountCapturableInMajorUnits)
```
Comment (kanikabansal08) [inline src/screens/Transaction/Order/OrderCaptureForm.res]: does the name `initiateValueJson` makes sense here?

File: src/screens/Transaction/Order/ShowOrder.res [hunk 12]
```diff
@@ -68,6 +69,14 @@ module ShowOrderDetails = {
               ? Normal
               : Disabled}
           />
+          <RenderIf condition={typedPaymentStatus === RequiresCapture && !(paymentId->isTestData)}>
```
Comment (kanikabansal08) [inline src/screens/Transaction/Order/ShowOrder.res]: can this button show for v2 page?
since this is not available for v2. should we add a check?

Comment (Riddhiagrawal001) [reply src/screens/Transaction/Order/ShowOrder.res]: As discussed with Deepanshu , this is present for v2 as well but we will restrict for only v1 version

File: src/screens/Transaction/Order/OrderCaptureForm.res (hunk 10 above)
Comment (Riddhiagrawal001) [reply src/screens/Transaction/Order/OrderCaptureForm.res:49]: It sends back normal payment response

**Review by susrithasabbini — DISMISSED** [review]

**Review by kanikabansal08 — DISMISSED** [review]

**Comment by kanikabansal08** [comment]
https://github.com/user-attachments/assets/a989c6fa-af55-4a5e-b41a-e629f79279e6

**Review by kanikabansal08 — APPROVED** [review]

**Review by susrithasabbini — APPROVED** [review]
